ppl/zqd/ingest/multipartreader.go

```diff
@@ -0,0 +1,129 @@
+package ingest
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"sync/atomic"
+
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/detector"
+	"github.com/brimsec/zq/zio/ndjsonio"
+	"github.com/brimsec/zq/zio/zngio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zqe"
+)
+
+type MulitpartLogReader struct {
+	mr       *multipart.Reader
+	opts     zio.ReaderOpts
+	stopErr  bool
+	warnings []string
+	zreader  *zbuf.File
+	zctx     *resolver.Context
+	nread    int64
+	done     bool
+}
+
+func NewMultipartLogReader(mr *multipart.Reader, zctx *resolver.Context) *MulitpartLogReader {
+	return &MulitpartLogReader{
+		mr:   mr,
+		opts: zio.ReaderOpts{Zng: zngio.ReaderOpts{Validate: true}},
+		zctx: zctx,
+	}
+}
+
+func (m *MulitpartLogReader) SetStopOnError() {
+	m.stopErr = true
+}
+
+func (m *MulitpartLogReader) Read() (*zng.Record, error) {
+read:
+	if m.done {
+		return nil, nil
+	}
+	if m.zreader == nil {
+		zr, err := m.next()
```

Comment:
A suggestion on handling the eof's: make the contract for `m.next()` that it always returns nil, nil when there's nothing else to read. Inside `MulitpartLogReader.next`, have it check if m.mr is nil, and if not, calls m.mr.NextPart(). If that errors with EOF, set m.mr to nil.

That will move all of the multipart logic handling inside next(), and remove the need for `done`.

Reply:
Much better, thanks